bson/vector.go

```diff
@@ -0,0 +1,271 @@
+	default:
+		return "invalid"
+	}
+}
+
+// These are vector conversion errors.
+var (
+	errInsufficientVectorData = errors.New("insufficient data")
+	errNonZeroVectorPadding   = errors.New("padding must be 0")
+	errVectorPaddingTooLarge  = errors.New("padding larger than 7")
+)
+
+type vectorTypeError struct {
+	Method string
+	Type   VectorDType
+}
+
+// Error implements the error interface.
+func (vte vectorTypeError) Error() string {
+	return "Call of " + vte.Method + " on " + vte.Type.String() + " vector"
+}
+
+// Vector represents a densely packed array of numbers / bits.
+type Vector struct {
+	dType       VectorDType
+	int8Data    []int8
+	float32Data []float32
+	bitData     []byte
+	bitPadding  uint8
+}
+
+// Type returns the vector type.
+func (v Vector) Type() VectorDType {
+	return v.dType
+}
+
+// Int8 returns the int8 slice hold by the vector.
+// It panics if v is not an int8 vector.
+func (v Vector) Int8() []int8 {
+	d, ok := v.Int8OK()
+	if !ok {
+		panic(vectorTypeError{"bson.Vector.Int8", v.dType})
+	}
+	return d
+}
+
+// Int8OK is the same as Int8, but returns a boolean instead of panicking.
+func (v Vector) Int8OK() ([]int8, bool) {
+	if v.dType != Int8Vector {
+		return nil, false
+	}
+	return v.int8Data, true
+}
+
+// Float32 returns the float32 slice hold by the vector.
+// It panics if v is not a float32 vector.
+func (v Vector) Float32() []float32 {
+	d, ok := v.Float32OK()
+	if !ok {
+		panic(vectorTypeError{"bson.Vector.Float32", v.dType})
+	}
+	return d
+}
+
+// Float32OK is the same as Float32, but returns a boolean instead of panicking.
+func (v Vector) Float32OK() ([]float32, bool) {
+	if v.dType != Float32Vector {
+		return nil, false
+	}
+	return v.float32Data, true
+}
+
+// PackedBit returns the byte slice representing the binary quantized (packed bit) vector and the byte padding, which
+// is the number of bits in the final byte that are to be ignored.
+// It panics if v is not a packed bit vector.
+func (v Vector) PackedBit() ([]byte, uint8) {
+	d, p, ok := v.PackedBitOK()
+	if !ok {
+		panic(vectorTypeError{"bson.Vector.PackedBit", v.dType})
+	}
+	return d, p
+}
+
+// PackedBitOK is the same as PackedBit, but returns a boolean instead of panicking.
+func (v Vector) PackedBitOK() ([]byte, uint8, bool) {
+	if v.dType != PackedBitVector {
+		return nil, 0, false
+	}
+	return v.bitData, v.bitPadding, true
+}
+
+// Binary returns the BSON Binary of the Vector.
+func (v Vector) Binary() Binary {
+	switch v.Type() {
+	case Int8Vector:
+		return binaryFromInt8Vector(v.Int8())
+	case Float32Vector:
+		return binaryFromFloat32Vector(v.Float32())
+	case PackedBitVector:
+		return binaryFromBitVector(v.PackedBit())
+	default:
+		panic("invalid Vector type")
+	}
+}
+
+func binaryFromInt8Vector(v []int8) Binary {
+	data := make([]byte, 2, len(v)+2)
+	copy(data, []byte{byte(Int8Vector), 0})
+	for _, e := range v {
+		data = append(data, byte(e))
+	}
+
+	return Binary{
+		Subtype: TypeBinaryVector,
+		Data:    data,
+	}
```

Comment:
Explicitly setting the size and filling it in reads more clearly IMO. Additionally, IIUC the following is slightly more efficient since we use a single allocation and avoid calling append in a loop:

```go
	data := make([]byte, len(v)+2)

	data[0] = byte(Int8Vector)
	data[1] = 0

	for i, e := range v {
		data[i+2] = byte(e)
	}

	return Binary{Subtype: TypeBinaryVector, Data: data}
```